src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs

```diff
@@ -85,10 +84,18 @@ public static OrderedSet<Symbol> Analyze(CSharpCompilation compilation, MethodSy
                     }
                     else
                     {
+                        var (type, refKindPrefix) = variable switch
+                        {
+                            LocalSymbol l => (l.Type, l.RefKind.ToLocalPrefix()),
+                            ParameterSymbol p => (p.Type, p.RefKind.ToParameterPrefix()),
+                            _ => throw ExceptionUtilities.UnexpectedValue(variable),
+                        };
+
                         foreach (CSharpSyntaxNode syntax in kvp.Value)
                         {
-                            // CS4013: Instance of type '{0}' cannot be used inside an anonymous function, query expression, iterator block or async method
-                            diagnostics.Add(ErrorCode.ERR_SpecialByRefInLambda, syntax.Location, type);
+                            // PROTOTYPE: Improve error message (the instance might be usable in iterator/async just not across yield/await)
+                            // CS4013: Instance of type '{0}{1}' cannot be used inside an anonymous function, query expression, iterator block or async method
```

Comment:
>// CS4013: Instance of type '{0}{1}' cannot be used inside an anonymous function, query expression, iterator block or async method [](http://example.com/codeflow?start=28&length=130)

I think an error referring to a specific local/parameter will be more informative for a user. Synthesized locals added during lowering might need special treatment though.

Reply:
I suggest addressing the issue with diagnostic instead of adding a  PROTOTYPE comment. And I suggest adding new errors for specific scenarios instead of modifying/reusing ```ERR_SpecialByRefInLambda```. 
I think distinct errors should be reported for the following scenarios:
- user declared local or parameter, report an error about the local or parameter (not about their type)
- lowering `ref` local 
- lowering local of restricted type

Reply:
The wording for new errors probably should be something like: " ...  cannot be preserved across 'await' or 'yield' boundary." Similar to ```ERR_RefReturningCallAndAwait```